lib/timeline-window.js

```diff
@@ -0,0 +1,381 @@
+        // not yet loaded
+        return [];
+    }
+
+    var result = [];
+
+    for (var timeline = this._start.timeline; timeline !== null;
+        timeline = timeline.getNeighbouringTimeline(false)) {
+
+        var events = timeline.getEvents();
+        var startIndex = 0, endIndex = events.length;
+        if (timeline === this._start.timeline) {
+            startIndex = this._start.index + timeline.getBaseIndex();
+        }
+        if (timeline === this._end.timeline) {
+            endIndex = this._end.index + timeline.getBaseIndex();
+        }
+
+        for (var i = startIndex; i < endIndex; i++) {
+            result.push(events[i]);
+        }
+    }
+    return result;
+};
+
+
+/**
+ * a thing which contains a timeline reference, and an index into it.
+ *
+ * @constructor
+ * @param {EventTimeline} timeline
+ * @param {number} index
+ * @private
+ */
+function TimelineIndex(timeline, index) {
+    this.timeline = timeline;
+
+    // the indexes are relative to BaseIndex, so could well be negative.
+    this.index = index;
+}
+
+/**
+ * @return {number} the minimum possible value for the index in the current
+ *    timeline
+ */
+TimelineIndex.prototype.minIndex = function() {
+    return -this.timeline.getBaseIndex();
+};
+
+/**
+ * @return {number} the maximum possible value for the index in the current
+ *    timeline
+ */
+TimelineIndex.prototype.maxIndex = function() {
+    return this.timeline.getEvents().length - this.timeline.getBaseIndex();
+};
+
+/**
+ * Try move the index forward, or into the neighbouring timeline
+ *
+ * @param {number} delta  number of events to advance by
+ * @return {number} number of events successfully advanced by
+ */
+TimelineIndex.prototype.advance = function(delta) {
+    // first try moving the cap
+    var x;
+    if (delta === 0) {
+        return 0;
+    } else if (delta < 0) {
+        x = Math.max(delta, this.minIndex() - this.index);
+        if (x < 0) {
+            this.index += x;
+            return x;
+        }
+    } else {
+        x = Math.min(delta, this.maxIndex() - this.index);
+        if (x > 0) {
+            this.index += x;
+            return x;
+        }
+    }
+
+    // next see if there is a neighbouring timeline to switch to
+    var neighbour = this.timeline.getNeighbouringTimeline(delta < 0);
+    if (neighbour) {
+        this.timeline = neighbour;
+        if (delta < 0) {
+            this.index = this.maxIndex();
+        } else {
+            this.index = this.minIndex();
+        }
+
+        debuglog("paginate: switched to new neighbour");
+
+        // recurse, using the next timeline
+        return this.advance(delta);
+    }
+
+    return 0;
+};
+
+/**
+ * Try move the index backwards, or into the neighbouring timeline
+ *
+ * @param {number} delta  number of events to retreat by
+ * @return {number} number of events successfully retreated by
+ */
+TimelineIndex.prototype.retreat = function(delta) {
+    return -this.advance(-delta);
```

Comment:
Spurious `-`? If this is supposed to return negative numbers then please:
 - Explicitly state this in the `@return` docs (rather than "the number of events")
 - Use `* -1` instead of `-`. The reason for this is to make it more obvious in the code that an inversion is occurring - I almost missed seeing the `-` because I did not expect to see it there.